### ISSUE TITLE
Unreachable Code Analyzer incorrectly alerts with catch when is in a for loop

### ISSUE THREAD

**Issue by mgaffigan** [dotnet/roslyn#41538]
#### Analyzer package

Not sure.  Given warning CS0162, I presume it to be an analyzer inside the compiler.

#### Package Version

Microsoft Visual Studio Enterprise 2019
Version 16.4.3
VisualStudio.16.Release/16.4.3+29709.97
Microsoft .NET Framework
Version 4.8.03752

Installed Version: Enterprise

C# Tools   3.4.1-beta4-19610-02+c4e5d138903b899477649a17f197abd2bcb22f9e

#### Diagnostic ID

[CS0162: Unreachable Code Detected](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0162)

#### Repro steps

1. Place try/catch with `when` inside for loop

```
    static void Main(string[] args)
    {
        for (int i = 0; i < 10; i++)
        //                      ^^^ Warning CS0162 Unreachable code detected
        {
            try
            {
                break;
            }
            catch when (i < 9)
            {
                throw;
            }
        }
        Console.WriteLine("Hello");
    }
```

2. Intellisense underline occurs for `i++` in `for` with CS0162 Unreachable Code

#### Expected behavior

No warning is given

#### Actual behavior

Warning is given.

**Comment by YairHalberstadt** [comment]
I'll take a stab at this

**Comment by YairHalberstadt** [comment]
@mgaffigan 

I'm not sure I understand. This looks unreachable to me!

**Comment by sharwell** [comment]
@mgaffigan I'm guessing this code was originally written thinking the `break` will exit the `try` block. However, in C# the `break` statement isn't related to `try`/`catch`/`finally` blocks, so for the code above it will exit the `for` loop. This means there is no path to the increment portion of the loop, and CS0162 is reported.

**Comment by CyrusNajmabadi** [comment]
> I'm not sure I understand. This looks unreachable to me!

Agreed.  Looks unreachable to me as well.

**Comment by mgaffigan** [comment]
Sorry folks.  In an attempt to get to an MCVE, I ended up breaking my example.  Here's the corrected code (note a conditional throw):

```
    static void Main(string[] args)
    {
        for (int i = 0; i < 10; i++)
        //                      ^^^ Warning CS0162 Unreachable code detected
        {
            try
            {
                if (DateTime.Now.Minute == 0)
                {
                    throw new Exception();
                }

                break;
            }
            catch when (i < 9)
            {
                throw;
            }
        }
        Console.WriteLine("Hello");
    }
```

Where I ran into this was looping over the results of `Dns.GetHostAddresses` and trying to connect to each one.  The first result which doesn't throw should break the loop, all but the last should ignore exceptions.
